### PR TITLE
fix(auth/interceptor): evitar logout indevido em 401 sem indicio de token invalido

### LINKED ISSUE #83: fix(auth/interceptor): evitar logout indevido ao receber 401 em rotas que não exigem reautenticação
## Contexto / Problema

Na tela `/pacientes/:id/sessoes`, ao clicar nos botões **Realizar** ou **Cancelar**, o usuário é deslogado e redirecionado para `/login`. Investigação mostrou que:

- O frontend chama `PATCH /api/sessoes/{id}/realizar` e `PATCH /api/sessoes/{id}/cancelar` (`src/app/core/services/sessao.service.ts:70-78`).
- Esses endpoints ainda não existem no backend (issue correspondente: `carlessopilatesapi`).
- Como o Spring Security retorna **401** para rotas inexistentes em vez de 404, o `auth.interceptor.ts` interpreta como sessão expirada e dispara `authService.logout()`.

A correção principal está no backend (implementar os endpoints + tratar 404 corretamente). Esta issue cobre a parte do frontend: tornar o interceptor mais defensivo para que falhas pontuais de 401 em endpoints específicos não derrubem a sessão inteira do usuário, e ajustar o serviço de sessões caso o backend mantenha o contrato `PATCH /{id}/realizar` | `/cancelar`.

## Objetivo

1. Confirmar/alinhar o contrato com o backend após implementação dos endpoints `realizar`/`cancelar`.
2. Tornar o tratamento de `401` no interceptor mais resiliente: só deslogar quando o token realmente é inválido/expirado, não em qualquer 401 vindo da API.
3. Melhorar a UX exibindo mensagem de erro específica em vez de logout silencioso.

### 1. `core/interceptors/auth.interceptor.ts`

- Em vez de chamar `authService.logout()` para qualquer `401`, distinguir os casos:
  - Se a resposta veio do endpoint de login → propagar o erro normalmente (já tratado por `isLoginRequest`).
  - Se o `401` for acompanhado de um corpo/cabeçalho que indique token inválido/expirado (ex.: `WWW-Authenticate: Bearer error=\"invalid_token\"` ou payload do backend com `code: TOKEN_EXPIRED`), aí sim fazer logout.
  - Caso contrário, **não** fazer logout: apenas propagar o erro para o componente decidir (mostrar toast/mensagem).
- Manter a sessão local intacta quando o 401 não for de autenticação real.

### 2. `core/services/sessao.service.ts` (alinhar com backend)

- Confirmar que os métodos `realizar(id)` e `cancelar(id)` continuam usando `PATCH /api/sessoes/{id}/realizar` e `/cancelar` após o backend implementar os endpoints.
- Caso o backend opte por outro contrato, ajustar o service e os testes (`sessao.service.spec.ts`).

### 3. `pages/pacientes/paciente-sessao-list/paciente-sessao-list.component.ts`

- O bloco `error` em `executarAcao()` já seta `this.erro`. Validar que, com o interceptor ajustado, a mensagem de erro fica visível (`this.erro = \`Erro ao ${acao} sessão.\``) em vez de provocar redirect para login.
- Considerar diferenciar erros 404/422 (backend pode retornar 422 para transição inválida) com mensagens específicas.

## Critérios de Aceite

- [ ] Após o backend disponibilizar os endpoints, clicar em **Realizar** muda o status da sessão para `REALIZADA` sem deslogar o usuário e exibe mensagem de sucesso.
- [ ] Clicar em **Cancelar** muda o status para `CANCELADA` sem deslogar o usuário e exibe mensagem de sucesso.
- [ ] Caso o backend retorne 401 em uma chamada que não é de autenticação (ex.: rota inexistente, falha pontual), o usuário **não** é deslogado; o componente exibe a mensagem de erro.
- [ ] Token realmente expirado/ inválido continua disparando logout e redirecionamento para `/login` (comportamento atual preservado).
- [ ] Testes unitários do interceptor e do `paciente-sessao-list.component` cobrem os novos cenários.

### Testes unitários
- `auth.interceptor.spec.ts`:
  - 401 com indicador de token inválido → chama `logout()`.
  - 401 sem indicador (ex.: erro de rota) → **não** chama `logout()`, propaga o erro.
  - 401 em request de login → propaga, sem logout.
- `paciente-sessao-list.component.spec.ts`:
  - Sucesso em `realizar`/`cancelar` → recarrega lista e mostra mensagem de sucesso.
  - Erro 4xx em `realizar`/`cancelar` → mostra mensagem de erro, **não** redireciona.
- `sessao.service.spec.ts`: confirmar URLs e método HTTP.

### Teste manual
- Logar como `admin`, abrir `/pacientes/{id}/sessoes`, clicar em **Realizar** e em **Cancelar** em sessões `AGENDADA`. A sessão deve mudar de status sem logout.
- Forçar token expirado (limpar/expirar `accessToken`) e fazer uma ação → redireciona para login (comportamento esperado).

## Observações Técnicas

- Esta issue **depende** da issue de backend (`carlessopilatesapi`: implementar `PATCH /sessoes/{id}/realizar` e `/cancelar` + tratar 404 de rota inexistente).
- Padrões do projeto: standalone components, Reactive Forms, services com `providedIn: 'root'`. Manter a convenção.
- Atualizar `docs/context.md` se houver mudança de contrato no service.

## Labels Sugeridas

`bug`, `frontend`, `enhancement`, `auth`

### PR DESCRIPTION
**Pull request by FabioCarlesso**
## Resumo
- `authInterceptor` agora só executa `logout()` quando a resposta `401` traz indicação explícita de token inválido/expirado: cabeçalho `WWW-Authenticate: Bearer error="invalid_token"` (ou `expired_token`) **ou** corpo com `code` em `{TOKEN_EXPIRED, INVALID_TOKEN, TOKEN_INVALID}`, e há sessão ativa.
- Demais `401` (ex.: rota inexistente que o Spring Security retorna como `401`, falhas pontuais do backend) são propagados ao componente sem derrubar a sessão.
- Mantido o comportamento atual para `401` no endpoint de login (apenas propaga o erro) e para token realmente expirado (logout + redirect para `/login`).

## Motivo
Na tela `/pacientes/:id/sessoes`, ao clicar em **Realizar** ou **Cancelar**, o frontend chamava endpoints ainda não implementados no backend (`PATCH /api/sessoes/{id}/realizar` e `/cancelar`). O Spring Security responde `401` para rotas inexistentes, e o interceptor anterior interpretava qualquer `401` fora do login como token expirado, derrubando a sessão do usuário. A correção torna o tratamento mais defensivo, alinhado ao escopo descrito na issue #83.

## Arquivos principais alterados
- `src/app/core/interceptors/auth.interceptor.ts` — checa `WWW-Authenticate` e `code` do corpo antes de chamar `logout()`; só age quando há sessão.
- `src/app/core/interceptors/auth.interceptor.spec.ts` — novos cenários: 401 com header `invalid_token`, 401 com `code: TOKEN_EXPIRED`, 401 genérico (sem dica) e 401 sem sessão local; mantidos cenários existentes.
- `docs/context.md` — registra o item 41 (correção do interceptor) e atualiza o trecho técnico de autenticação.
- `README.md` — atualiza a descrição do interceptor JWT.

## Testes executados
- `npx ng test --watch=false --browsers=ChromeHeadless` → **540 SUCCESS**.
- `npm run build` → bundle de produção gerado sem erros.

## Referência
Closes #83

### CODE REVIEW

File: src/app/core/interceptors/auth.interceptor.ts [hunk 1]
```diff
@@ -16,11 +19,37 @@ export const authInterceptor: HttpInterceptorFn = (req, next) => {
...
       return throwError(() => error);
     })
   );
 };
+
+function isTokenInvalidResponse(error: HttpErrorResponse): boolean {
+  const wwwAuthenticate = error.headers?.get('WWW-Authenticate') ?? '';
+  if (wwwAuthenticate) {
+    const lowered = wwwAuthenticate.toLowerCase();
+    if (TOKEN_INVALID_HEADER_HINTS.some(hint => lowered.includes(hint))) {
```
Comment (FabioCarlesso) [inline src/app/core/interceptors/auth.interceptor.ts]: Problem:
The header check treats any occurrence of `invalid_token` or `expired_token` anywhere in `WWW-Authenticate` as proof that the token is invalid.

Why it matters:
This can still produce the same class of false logout this PR is trying to avoid. For example, a `401` with `Bearer error="invalid_request", error_description="... invalid_token ..."` or another diagnostic string containing that text would clear the session even though the Bearer `error` value is not one of the token-expiration cases.

Suggested improvement:
Parse the `WWW-Authenticate` Bearer challenge and compare the `error` parameter value exactly, e.g. only logout when `error` is `invalid_token` or `expired_token`. Add a negative spec where the header contains `invalid_token` in another parameter/description but `error` is not token-invalid.

Severity: Medium

File: docs/context.md [hunk 2]
```diff
@@ -52,6 +52,7 @@ O projeto está em fase inicial de desenvolvimento (**MVP**). Commits realizados
...
...gamento, detalhamento por aula realizada e exportação em PDF/XLSX, além do relatório fiscal de emissão de NFSEs por competência com exportação CSV/XLSX. A seção administrativa em `/admin` está disponível como hub inicial protegido por `roleGuard(['ADMIN'])`, e a listagem `/admin/usuarios` já consome o `UsuarioAdminService` com paginação server-side, exibe nome, e-mail, perfil e status, oferece criar/editar/inativar/reativar/excluir com confirmação antes de ações destrutivas e bloqueia o usuário logado de inativar ou excluir a própria conta. O `UsuarioAdminService` fornece a camada de serviço tipada para os endpoints `/api/users`, cobrindo listagem paginada, busca por ID, busca do perfil próprio, criação, atualização, exclusão, ativação, inativação e consulta de roles disponíveis. A aplicação agora pode ser executada em container Docker. A autenticação via JWT está implementada: tela de login, guard de rotas, interceptor HTTP, logout e redirecionamento para login quando uma chamada autenticada retorna `401`. O estilo global do Angular consome tokens do Design System Carlesso em `src/styles/_tokens.scss`, com estratégia de tema e densidade centralizada pelo `StylePreferencesService`.
```
Comment (FabioCarlesso) [inline docs/context.md]: Problem:
This summary still says authentication redirects on `401` for authenticated calls, which conflicts with the new behavior documented just above and implemented by the interceptor.

Why it matters:
`context.md` is used as project memory, so this stale sentence can lead future changes back toward the old broad-logout behavior and make the auth contract ambiguous.

Suggested improvement:
Update this sentence to say logout/redirection only happens when a `401` explicitly indicates an invalid or expired token, while other `401` responses are propagated to the component.

Severity: Low

**Review by FabioCarlesso — COMMENTED** [review]
Review completed.

Areas that look good: API request/response mapping, loading/error UI behavior, route guard integration, token storage behavior, RxJS usage, performance impact, and the focused interceptor tests are appropriate for this PR. I also ran `npx ng test --watch=false --browsers=ChromeHeadless --include='src/app/core/interceptors/auth.interceptor.spec.ts'` locally and the interceptor suite passed with 8 specs.

I left two findings below: one edge-case bug in the header parsing and one documentation inconsistency.

File: src/app/core/interceptors/auth.interceptor.ts (hunk 1 above)
Comment (FabioCarlesso) [reply src/app/core/interceptors/auth.interceptor.ts]: Corrigido no commit 05692f1. O interceptor agora extrai o parâmetro `error` do desafio `Bearer` e compara o valor exatamente contra `invalid_token`/`expired_token`, em vez de procurar esses textos em todo o header. Também adicionei teste negativo para `invalid_token` em `error_description` com `error="invalid_request"`.

File: docs/context.md (hunk 2 above)
Comment (FabioCarlesso) [reply docs/context.md]: Corrigido no commit 05692f1. Atualizei a frase de resumo em `docs/context.md` para deixar claro que o redirecionamento para `/login` só ocorre quando o `401` indica token inválido/expirado, e que os demais `401` são propagados ao componente.
